s3/s3_operations.cpp

```diff
@@ -845,6 +849,55 @@ namespace irods_s3 {
 
             irods::error result = SUCCESS();
 
+            irods::file_object_ptr file_obj = boost::dynamic_pointer_cast<irods::file_object>(_ctx.fco());
+
+            // create entry for fd in per_thread_data if it doesn't exist
+            int fd = file_obj->file_descriptor();
+            per_thread_data data{};
+            if (fd_data.exists(fd)) {
+                data = fd_data.get(fd);
+            }
+
+            // Issue 2124 - On a PUT operation with checksum, the open_mode was forced to O_CREAT, O_WRONLY, O_TRUNC so that streaming
+            // would work.  We need to check if this was a PUT_OPR and if so we have moved on to the read-only portion of
+            // processing for checksum calculation.  If that is the case, the open_mode will be reset to O_RDONLY.
+            // So that this isn't done on every read call, a flag (checked_for_checksum_read_after_put) is set to indicate that we
+            // have already checked for this condition and adjusted the open_mode accordingly.
+            if (!data.checked_for_checksum_read_after_put) {
+
+                // Get oprType
+                int oprType = -1;
+                bool found = false;
+                for (int i = 0; i < NUM_L1_DESC; ++i) {
+                   if (L1desc[i].inuseFlag) {
+                       if (L1desc[i].dataObjInp && L1desc[i].dataObjInfo &&
+                               L1desc[i].dataObjInp->objPath == file_obj->logical_path()
+                               && L1desc[i].dataObjInfo->filePath == file_obj->physical_path()) {
+
+                           found = true;
+                           oprType = L1desc[i].dataObjInp->oprType;
+                       }
+                   } else if (found) {
+                       break;
+                   }
+                }
+                
+                data.checked_for_checksum_read_after_put = true;
+
+                // If oprType is PUT_OPR, adjust the open mode to O_RDONLY.
+                if (oprType == PUT_OPR) {
```

Comment:
Is it a problem if `oprType == -1` at this point? That might mean that the L1 descriptor is invalid, right?

Reply:
I think it would make sense to continue since this check is a fix for a specific scenario meaning if we encounter anything else we just proceed normally.

That may be an indication of a problem but it isn't related to what we are doing here.